pkgs/development/go-packages/generic/default.nix

```diff
@@ -48,26 +48,56 @@ let
 
   removeExpr = refs: ''remove-references-to ${lib.concatMapStrings (ref: " -t ${ref}") refs}'';
 
+  fetchSrc = fetch:
+    if fetch.type == "git" then
+      fetchgit {
+        inherit (fetch) url rev sha256;
+      }
+    else if fetch.type == "hg" then
+      fetchhg {
+        inherit (fetch) url rev sha256;
+      }
+    else if fetch.type == "bzr" then
+      fetchbzr {
+        inherit (fetch) url rev sha256;
+      }
+    else if fetch.type == "FromGitHub" then
+      fetchFromGitHub {
+        inherit (fetch) owner repo rev sha256;
+      }
+    else abort "Unrecognized package fetch type: ${fetch.type}";
+
   dep2src = goDep:
-    {
+    let
+      otherMajorVersions = map (v: {
+        src = fetchsrc v.fetch;
+        inherit (v) majorVersion;
+      }) (goDep.otherMajorVersions or []);
+
+      main = fetchSrc goDep.fetch;
+
+      combined = stdenv.mkDerivation {
+        inherit (main) name;
+
+        unpackPhase = ''
+          unpackFile "${main}"
+        '' + lib.flip lib.concatMapStrings otherMajorVersions (drv: ''
+          mkdir ./${drv.majorVersion}
+          pushd ./${drv.majorVersion}
```

Comment:
Sorry I just took the liberty to push my untested changes, since I don't have an example to test this and the time to find one. But this variant seems to be closer to your intention. If not, feel free to revert or adapt the last commit that I made.